state/protocol/badger/validity.go

```diff
@@ -197,6 +200,58 @@ func isValidEpochCommit(commit *flow.EpochCommit, setup *flow.EpochSetup) error
 	return nil
 }
 
+// isValidVersionBeacon validates internal structure of flow.VersionBeacon
+// protocol.InvalidServiceEventError with appropriate message is returned if any validation fails.
+// Please refer to documentation of flow.VersionBeacon for a description of what's expected of a valid VersionBeacon.
+func isValidVersionBeacon(vb *flow.VersionBeacon) error {
+	if len(vb.RequiredVersions) == 0 {
+		return protocol.NewInvalidServiceEventError("required versions empty")
+	}
+
+	// handle case when only one version is present
+	if len(vb.RequiredVersions) == 1 {
+		_, err := validateRequirement(vb.RequiredVersions[0])
+		return err
+	}
+
+	for i := 0; i < len(vb.RequiredVersions)-1; i++ {
+		current := vb.RequiredVersions[i]
+		next := vb.RequiredVersions[i+1]
+
+		// next version must higher than previous one
+		if current.Height >= next.Height {
+			return protocol.NewInvalidServiceEventError("higher requirement (index=%d) height(%d) at or below previous height(%d) (index=%d)", i+1, next.Height, current.Height, i)
+		}
+
+		currentVersion, err := validateRequirement(current)
+		if err != nil {
+			return err
+		}
+
+		nextVersion, err := validateRequirement(next)
+		if err != nil {
+			return err
+		}
+
+		if nextVersion.LessThan(*currentVersion) {
```

Comment:
So we allow current and next to be equal?

Reply:
Yes. It doesn't create any confusion (should basically be a no-op)